### DISCUSSION TITLE
UPBGE 0.3 - Feature request: Copying and pasting logic bricks

### DISCUSSION THREAD

**Discussion by arch-sketch** [UPBGE/upbge#1465] [Ideas/Features for UPBGE 0.3+]
To work with logic bricks and manage them, it is important to be able to separate different logic trees in state slots. But because of the lack of a copy/paste functionality, it becomes very laborious to recreate logic trees from scratch for every new "layer", specially when talking about more complex ones, or ones that are identical apart from the values inside each brick. It would also be useful even for simple tasks like duplicating a logic brick with all the options already set (type, mode, objects, values, etc) and then just change what is needed, so you wont have to add a new one and start from zero each time.

A copy/paste function would also be crucial when copying logic brick trees between objects, allowing you to add logic to objects without replacing the logic they already contain (which is what happens when you use the object - game - copy logic bricks function).

So the feature request would be to be able to select, copy and paste logic bricks inside or between states in a single object, and also between different objects, while preserving their connections, if they are indeed connected.

**Comment by pozitigor1** [comment]
Wow, I'm not only one who think that copying/pasting logic bricks would increase work perfomance. But I also must admit that there's no way to make one logic brick tree available for multiple objects. So it would be better to create new window in logic brick tree for that things.
Sure, I could done it by myself but I'm novice in upbge and python :p

**Reply by arch-sketch** [reply]
Indeed, logic trees are designed to be unique for each object and thats great like that. What I meant is that if a copy/paste function existed, we would be able to easily copy logic or parts of it from one object to another, in a very convenient way (you can copy logic between objects right now, but the copied logic is not added to the object, it fully replaces whatever was there before). Freely copying/pasting logic bricks would be such a time saver, I have no idea why they didnt implement this basic functionality when creating the system originally... It makes no sense. Maybe theres a strong reason, coding wise. Still, there's hope we might get this feature in the future with the bricky nodes addon, I suggested this to IzaZed, the addon creater, and he agreed to work on it.
